src/module.c

```diff
@@ -10846,36 +10880,27 @@ NULL
             addReplyError(c,
                 "Error loading the extension. Please check the server logs.");
     } else if (!strcasecmp(subcmd,"unload") && c->argc == 3) {
-        if (moduleUnload(c->argv[2]->ptr) == C_OK)
-            addReply(c,shared.ok);
-        else {
-            char *errmsg;
-            switch(errno) {
-            case ENOENT:
-                errmsg = "no such module with that name";
-                break;
-            case EBUSY:
-                errmsg = "the module exports one or more module-side data "
-                         "types, can't unload";
-                break;
-            case EPERM:
-                errmsg = "the module exports APIs used by other modules. "
-                         "Please unload them first and try again";
-                break;
-            case EAGAIN:
-                errmsg = "the module has blocked clients. "
-                         "Please wait them unblocked and try again";
-                break;
-            case EINPROGRESS:
-                errmsg = "the module holds timer that is not fired. "
-                         "Please stop the timer or wait until it fires.";
-                break;
-            default:
-                errmsg = "operation not possible.";
-                break;
+        const char *errmsg = NULL;
+        struct RedisModule *module = dictFetchValue(modules,c->argv[2]->ptr);
+
+        if (server.in_exec) {
+            if (!moduleShouldUnload(module,&errmsg)) {
+                addReplyErrorFormat(c,"Error unloading module: %s",errmsg);
+                return;
             }
-            addReplyErrorFormat(c,"Error unloading module: %s",errmsg);
+
+            /* Note: we need to unload the module asynchronously because this module command
+             * may already in MULTI command queue where the module can't be unloaded safely. */
+            module->unload_asap = 1;
+            addReplyStatus(c,"Module unload command called from transaction, "
+                           "module will be unloaded when the transaction ends");
```

Comment:
Status (simple string) are usually short and clients can match them to check the result. Can we return "QUEUED", "ASYNC", "ASAP" or something like that here?